### ISSUE TITLE
(v9) Popover: need inline prop

### ISSUE THREAD

**Issue by smhigley** [microsoft/fluentui#21525]
## Background:

Popups that appear at the end of the DOM, away from their trigger, cause a number of accessibility issues, depending on other contextual factors:
- If the popup is non-modal, a screen reader user navigating via virtual cursor can exit the popup and will land at the end of the DOM
- If the popup doesn't manage focus into itself on open and back to the trigger on close (e.g. if the popup has no focusable items, or the trigger intentionally maintains focus), there's no good way to access the popup.
- Some components can't send focus to/from the popup, e.g. Combobox. For combobox specifically, the primary issue is with touch screen readers.
- For popups that don't have visible close buttons, touch screen reader users don't have any clear way to exit the popup.
- In pretty rare cases, there may be a non-modal popup where focus remains on the trigger when opened, but popup has focusable children (e.g. something similar to a tooltip, but with links or buttons).

## Potential solutions:
- Add a prop like `inline` so authors can choose to render a popup inline with the trigger in the DOM (this would be the minimum we would need to do, I think)
- Render popups inline by default (this would be the breaking API change)
- use `aria-modal="true"` by default for all non-inline dialog-like popups even if they're visibly non-modal, so virtual cursor navigation doesn't dump people at the end of the DOM
- Add a visually hidden close button that isn't in the tab order (or arrow order for menu/combo/etc) -- React Spectrum does this, so maybe not as weird as it seems on first glance
- if Popover doesn't have interactive children, perhaps we should send focus to the root by default (we'd need to figure out naming for this)
- scary docs about the accessibility implications of non-inline?

### Some considerations for default inline vs. portal:
- If inline is the default, we'll likely get teams frustrated about unexpected clipping/layering context issues.
- If portal is the default, we'll likely get teams frustrated about unexpected a11y issues after they have a test pass.
- One of the goals of a11y in vNext is eliminating random "gotcha" (not gacha) props that authors have to know and set to get the recommended experience
- Some of the unknowns include how frequently parent layer context is an issue for Fluent consumers, and how many of the a11y mitigations we can include

**Comment by jurokapsiar** [comment]
There are several layouting, positioning and sizing restrictions for inline popups. They get very tricky for example for larger zoom levels where the parents might be restricting the layout of children.

At least few weeks back (this might have changed), it was still required to set `aria-hidden="true"` to elements outside of the modal to achieve reliable restriction of navigation, even if aria-modal was used. It is relatively easy to set aria-hidden if the popup is not inline, however if it is inline, this might be a complex task.
From my experience, modal popups should not be inline.

There might be an option to use `aria-controls`, at lease JAWS allows users to jump to the controlled content: https://a11ysupport.io/tests/tech__aria__aria-controls, we can try to file bugs for other screen readers. However, phone devices will most probably not support this. Alternatively we can check if `aria-owns` might help to move the popup closer to the relevant content.

I think giving the control to the consumer is the right thing to do. There is no single solution that would work for all cases, so it will need to be responsibility of the consumer to choose.

**Comment by smhigley** [comment]
Resolution:

- Add an inline prop, make it false by default
- Add the close button and `aria-modal` for all non-inline popups (we can revisit the second if it causes issues in practice, since it wouldn't be a breaking change)
- Focus on solving the use case of a popover with no interactive elements, and maybe update the default example in our docs to not be that
- Needs good documentation on the implications inline vs. not

**Comment by smhigley** [comment]
@jurokapsiar I agree with everything you wrote, and I think we need really good docs for this to help authors decide when they need to set `inline`. Definitely not for modal popups 😅.

For the `aria-controls` thing, JAWS actually removed the announcement of a controlled element by default because of user feedback, so I don't think we'll get a lot of success trying to use that. `aria-owns` should help with Talkback and desktop screen reader virtual cursor navigation, so I definitely agree we should look at implementing it even if VoiceOver is still left out.

**Comment by smhigley** [comment]
Just a small note: we may not want to use the prop name `inline` for this, since that is already used for form controls as the block vs. inline CSS `display` variation. Alternatively, we may want to change the `inline` prop name on Input/Select/Combo to something else.

**Comment by ling1726** [comment]
Already fixed in #22789
